### PR TITLE
Fix `troch.gradient` check for spacing arg list length

### LINKED ISSUE #114207: `torch.gradient` heap buffer overflow
### 🐛 Describe the bug

`torch.gradient` causes heap buffer overflow with specific input tensor shape.

Test code:
```python
import torch
t = torch.randn(2,2)
coords = (torch.randn(2),)
result = torch.gradient(t, spacing=coords)
```
Error log:
```
=================================================================
==1432440==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020008afb98 at pc 0x7f843a254b6e bp 0x7ffcafae5e90 sp 0x7ffcafae5e88
READ of size 8 at 0x6020008afb98 thread T0
    #0 0x7f843a254b6d in c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>::operator->() const /home/sehoon/pytorch-latest-asan/c10/util/intrusive_ptr.h:412:12
    #1 0x7f843a27c434 in at::TensorBase::dim() const /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/TensorBase.h:118:12
    #2 0x7f843bf8a9a1 in at::native::gradient_helper(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::ArrayRef<long>, long) /home/sehoon/pytorch-latest-asan/aten/src/ATen/native/ReduceOps.cpp:1040:5
    #3 0x7f843bf9067c in at::native::gradient(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) /home/sehoon/pytorch-latest-asan/aten/src/ATen/native/ReduceOps.cpp:1121:10
    #4 0x7f844261a174 in at::(anonymous namespace)::(anonymous namespace)::wrapper_CompositeImplicitAutograd_tensorarrayint_gradient(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) /home/sehoon/pytorch-latest-asan/build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:2179:10
    #5 0x7f844261af9f in c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<std::vector<at::Tensor, std::allocator<at::Tensor> > (at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_CompositeImplicitAutograd_tensorarrayint_gradient(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long))>, std::vector<at::Tensor, std::allocator<at::Tensor> >, c10::guts::typelist::typelist<at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long> >::operator()(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:13:16
    #6 0x7f844261af9f in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<std::vector<at::Tensor, std::allocator<at::Tensor> > (at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long), &(at::(anonymous namespace)::(anonymous namespace)::wrapper_CompositeImplicitAutograd_tensorarrayint_gradient(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long))>, std::vector<at::Tensor, std::allocator<at::Tensor> >, c10::guts::typelist::typelist<at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long> >, std::vector<at::Tensor, std::allocator<at::Tensor> > (at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:468:14
    #7 0x7f843f3ad5fe in std::vector<at::Tensor, std::allocator<at::Tensor> > c10::callUnboxedKernelFunction<std::vector<at::Tensor, std::allocator<at::Tensor> >, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long>(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<at::Tensor>&&, c10::optional<long>&&, long&&) /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/boxing/KernelFunction_impl.h:50:12
    #8 0x7f843edb6aec in std::vector<at::Tensor, std::allocator<at::Tensor> > c10::KernelFunction::call<std::vector<at::Tensor, std::allocator<at::Tensor> >, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long>(c10::OperatorHandle const&, c10::DispatchKeySet, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) const /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/boxing/KernelFunction_impl.h:102:16
    #9 0x7f843edb6aec in std::vector<at::Tensor, std::allocator<at::Tensor> > c10::Dispatcher::call<std::vector<at::Tensor, std::allocator<at::Tensor> >, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long>(c10::TypedOperatorHandle<std::vector<at::Tensor, std::allocator<at::Tensor> > (at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long)> const&, at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) const /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/dispatch/Dispatcher.h:673:26
    #10 0x7f843edb6aec in c10::TypedOperatorHandle<std::vector<at::Tensor, std::allocator<at::Tensor> > (at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long)>::call(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) const /home/sehoon/pytorch-latest-asan/aten/src/ATen/core/dispatch/Dispatcher.h:521:41
    #11 0x7f843edb6aec in at::_ops::gradient_tensorarrayint::call(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) /home/sehoon/pytorch-latest-asan/build/aten/src/ATen/Operators_2.cpp:3009:15
    #12 0x7f8465d08014 in at::gradient(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) /home/sehoon/pytorch-latest-asan/build/aten/src/ATen/ops/gradient.h:52:12
    #13 0x7f8465cb4566 in torch::autograd::THPVariable_gradient(_object*, _object*, _object*)::$_105::operator()(at::Tensor const&, c10::ArrayRef<at::Tensor>, c10::optional<long>, long) const /home/sehoon/pytorch-latest-asan/torch/csrc/autograd/generated/python_torch_functions_1.cpp:3077:16
    #14 0x7f8465bb1b64 in torch::autograd::THPVariable_gradient(_object*, _object*, _object*) /home/sehoon/pytorch-latest-asan/torch/csrc/autograd/generated/python_torch_functions_1.cpp:3079:19
    #15 0x507386 in cfunction_call /usr/local/src/conda/python-3.9.18/Objects/methodobject.c:543:19
    #16 0x4f073b in _PyObject_MakeTpCall /usr/local/src/conda/python-3.9.18/Objects/call.c:191:18
    #17 0x4ece25 in _PyObject_VectorcallTstate /usr/local/src/conda/python-3.9.18/Include/cpython/abstract.h:116:16
    #18 0x4ece25 in _PyObject_VectorcallTstate /usr/local/src/conda/python-3.9.18/Include/cpython/abstract.h:103:1
    #19 0x4ece25 in PyObject_Vectorcall /usr/local/src/conda/python-3.9.18/Include/cpython/abstract.h:127:12
    #20 0x4ece25 in call_function /usr/local/src/conda/python-3.9.18/Python/ceval.c:5077:13
    #21 0x4ece25 in _PyEval_EvalFrameDefault /usr/local/src/conda/python-3.9.18/Python/ceval.c:3537:19
    #22 0x4e6b29 in _PyEval_EvalFrame /usr/local/src/conda/python-3.9.18/Include/internal/pycore_ceval.h:40:12
    #23 0x4e6b29 in _PyEval_EvalCode /usr/local/src/conda/python-3.9.18/Python/ceval.c:4329:14
    #24 0x4e67b6 in _PyEval_EvalCodeWithName /usr/local/src/conda/python-3.9.18/Python/ceval.c:4361:12
    #25 0x4e6768 in PyEval_EvalCodeEx /usr/local/src/conda/python-3.9.18/Python/ceval.c:4377:12
    #26 0x59466a in PyEval_EvalCode /usr/local/src/conda/python-3.9.18/Python/ceval.c:828:12
    #27 0x5c1dc6 in run_eval_code_obj /usr/local/src/conda/python-3.9.18/Python/pythonrun.c:1221:9
    #28 0x5bddcf in run_mod /usr/local/src/conda/python-3.9.18/Python/pythonrun.c:1242:19
    #29 0x45674d in pyrun_file.cold /usr/local/src/conda/python-3.9.18/Python/pythonrun.c:1140:15
    #30 0x5b7ab1 in pyrun_simple_file /usr/local/src/conda/python-3.9.18/Python/pythonrun.c:450:13
    #31 0x5b7ab1 in PyRun_SimpleFileExFlags /usr/local/src/conda/python-3.9.18/Python/pythonrun.c:483:15
    #32 0x5b502d in pymain_run_file /usr/local/src/conda/python-3.9.18/Modules/main.c:379:15
    #33 0x5b502d in pymain_run_python /usr/local/src/conda/python-3.9.18/Modules/main.c:604:21
    #34 0x5b502d in Py_RunMain /usr/local/src/conda/python-3.9.18/Modules/main.c:683:5
    #35 0x588718 in Py_BytesMain /usr/local/src/conda/python-3.9.18/Modules/main.c:1129:12
error: failed to decompress '.debug_aranges', zlib is not available
error: failed to decompress '.debug_info', zlib is not available
error: failed to decompress '.debug_abbrev', zlib is not available
error: failed to decompress '.debug_line', zlib is not available
error: failed to decompress '.debug_str', zlib is not available
error: failed to decompress '.debug_line_str', zlib is not available
error: failed to decompress '.debug_loclists', zlib is not available
error: failed to decompress '.debug_rnglists', zlib is not available
    #36 0x7f8475b1ad8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    #37 0x7f8475b1ae3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
    #38 0x5885cd in _start (/home/sehoon/anaconda3/envs/torch-latest-asan/bin/python3.9+0x5885cd)

0x6020008afb98 is located 0 bytes to the right of 8-byte region [0x6020008afb90,0x6020008afb98)
allocated by thread T0 here:
    #0 0x7f8475ecf0bd in operator new(unsigned long) /tmp/llvm/utils/release/final/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:99:3
    #1 0x7f8465b12c59 in __gnu_cxx::new_allocator<at::Tensor>::allocate(unsigned long, void const*) /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/ext/new_allocator.h:127:27
    #2 0x7f8465b12be0 in std::allocator_traits<std::allocator<at::Tensor> >::allocate(std::allocator<at::Tensor>&, unsigned long) /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/alloc_traits.h:464:20
    #3 0x7f8465b12b9f in std::_Vector_base<at::Tensor, std::allocator<at::Tensor> >::_M_allocate(unsigned long) /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:346:20
    #4 0x7f8465b12a60 in std::_Vector_base<at::Tensor, std::allocator<at::Tensor> >::_M_create_storage(unsigned long) /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:361:33
    #5 0x7f8465b126b4 in std::_Vector_base<at::Tensor, std::allocator<at::Tensor> >::_Vector_base(unsigned long, std::allocator<at::Tensor> const&) /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:305:9
    #6 0x7f8465b12318 in std::vector<at::Tensor, std::allocator<at::Tensor> >::vector(unsigned long, std::allocator<at::Tensor> const&) /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:511:9
    #7 0x7f8465b11731 in torch::PythonArgs::tensorlist(int) /home/sehoon/pytorch-latest-asan/torch/csrc/utils/python_arg_parser.h:448:27
    #8 0x7f8465bb19c6 in torch::autograd::THPVariable_gradient(_object*, _object*, _object*) /home/sehoon/pytorch-latest-asan/torch/csrc/autograd/generated/python_torch_functions_1.cpp:3079:54
    #9 0x507386 in cfunction_call /usr/local/src/conda/python-3.9.18/Objects/methodobject.c:543:19

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/sehoon/pytorch-latest-asan/c10/util/intrusive_ptr.h:412:12 in c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>::operator->() const
Shadow bytes around the buggy address:
  0x0c048010df20: fa fa fd fa fa fa fd fd fa fa fd fa fa fa fd fa
  0x0c048010df30: fa fa fd fa fa fa fd fa fa fa fd fd fa fa fd fa
  0x0c048010df40: fa fa fd fd fa fa fd fa fa fa fd fa fa fa fd fa
  0x0c048010df50: fa fa fd fd fa fa fd fa fa fa fd fd fa fa fd fa
  0x0c048010df60: fa fa fd fa fa fa fd fa fa fa fd fd fa fa fd fa
=>0x0c048010df70: fa fa 00[fa]fa fa 00 00 fa fa fd fd fa fa fd fa
  0x0c048010df80: fa fa fd fa fa fa fd fa fa fa fd fa fa fa fd fd
  0x0c048010df90: fa fa 00 fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c048010dfa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c048010dfb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c048010dfc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1432440==ABORTING
```

### Versions

PyTorch version: 2.2.0a0+git140c54e
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 12.0.1
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.9.18 (main, Sep 11 2023, 13:41:44)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-87-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             112
On-line CPU(s) list:                0-111
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6348 CPU @ 2.60GHz
CPU family:                         6
Model:                              106
Thread(s) per core:                 2
Core(s) per socket:                 28
Socket(s):                          2
Stepping:                           6
CPU max MHz:                        3500.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
L1d cache:                          2.6 MiB (56 instances)
L1i cache:                          1.8 MiB (56 instances)
L2 cache:                           70 MiB (56 instances)
L3 cache:                           84 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-27,56-83
NUMA node1 CPU(s):                  28-55,84-111
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] optree==0.10.0
[pip3] torch==2.2.0a0+git140c54e
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.1.0         h06a4308_46344  
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] torch                     2.2.0a0+git140c54e           dev_0    <develop>

cc @ezyang @gchanan @zou3519 @kadeng @albanD

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #114207

cc @malfet

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -1009,7 +1009,12 @@ static void pre_check_gradient(const Tensor& self, c10::optional<int64_t> spacin
   // Helper for gradient function to make sure input data satisfies prerequisites
   TORCH_CHECK(self.scalar_type() != ScalarType::Byte, "torch.gradient does not support uint8 input.");
   if (spacing_size.has_value() && !dim.has_value()) {
-    TORCH_CHECK(spacing_size.value() == 1 || spacing_size.value() == self.dim(), "torch.gradient expected spacing to be unspecified, a scalar or a list of length ", self.dim(), " but got a list of length ", spacing_size.value());
+    // NOTE: If spacing was given as a scalar, the callers of this function
+    // create a spacing vector of the expected size, and this check passes
```
Comment (albanD) [inline aten/src/ATen/native/ReduceOps.cpp:1013]: "spacing vector" here is "dim" right?

Comment (kurtamohler) [reply aten/src/ATen/native/ReduceOps.cpp:1013]: No I was referring to the "spacing" arg. Here's an example of one of the callers creating a vector when spacing (named `unit_size` in this impl) is given as a scalar:

https://github.com/pytorch/pytorch/blob/639060cb0b719a652edd74981fe98b927a314cab/aten/src/ATen/native/ReduceOps.cpp#L1144-L1148

Comment (albanD) [reply aten/src/ATen/native/ReduceOps.cpp:1013]: Thanks for the explanation!

**Review by albanD — APPROVED** [review]
Sounds good!
Thanks for fixing this!

**Pull request closed without merging**